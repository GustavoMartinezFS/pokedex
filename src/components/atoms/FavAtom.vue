<template>
  <div class="fav-button" @click="toggleFavorite">
    <i v-if="isFavorite">
      <img src="@/assets/star-gold.svg" alt="star-gold">
    </i>
    <i v-else>
      <img src="@/assets/star-gray.svg" alt="star">
    </i>
  </div>
</template>

<script lang="ts" setup>
import { defineProps, ref, watch } from 'vue';

const props = defineProps({
  isFavorite: Boolean,
});

const isFavorite = ref(props.isFavorite ?? false);

const emit = defineEmits(['update:isFavorite']);

const toggleFavorite = () => {
  emit('update:isFavorite', !isFavorite.value);
};

watch(() => props.isFavorite, (newVal) => {
  isFavorite.value = newVal;
});
</script>

<style scoped>
.fav-button {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: var(--c-dark-white);
}
</style>
