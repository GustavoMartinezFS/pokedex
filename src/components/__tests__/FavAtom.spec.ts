import { mount } from '@vue/test-utils';
import FavAtom from '@/components/atoms/FavAtom.vue';

describe('FavAtom.vue', () => {
  it('renders the correct icon based on isFavorite prop', async () => {
    const wrapper = mount(FavAtom, {
      props: {
        isFavorite: true,
      },
    });

    const starGold = wrapper.find('img[alt="star-gold"]');
    expect(starGold.exists()).toBe(true);

    await wrapper.setProps({ isFavorite: false });

    const starGray = wrapper.find('img[alt="star"]');
    expect(starGray.exists()).toBe(true);
  });

  it('emits update:isFavorite event when clicked', async () => {
    const wrapper = mount(FavAtom, {
      props: {
        isFavorite: false,
      },
    });

    await wrapper.find('.fav-button').trigger('click');
    expect(wrapper.emitted('update:isFavorite')).toBeTruthy();
    expect(wrapper.emitted('update:isFavorite')?.[0]).toEqual([true]);

    await wrapper.setProps({ isFavorite: true });

    await wrapper.find('.fav-button').trigger('click');
    expect(wrapper.emitted('update:isFavorite')?.[1]).toEqual([false]);
  });

  it('updates local isFavorite state when prop changes', async () => {
    const wrapper = mount(FavAtom, {
      props: {
        isFavorite: false,
      },
    });

    expect(wrapper.vm.isFavorite).toBe(false);

    await wrapper.setProps({ isFavorite: true });

    expect(wrapper.vm.isFavorite).toBe(true);
  });

  it('matches snapshot', () => {
    const wrapper = mount(FavAtom, {
      props: {
        isFavorite: true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
