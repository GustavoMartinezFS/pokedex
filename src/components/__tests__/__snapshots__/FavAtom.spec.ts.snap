// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavAtom.vue > matches snapshot 1`] = `"<div data-v-bd0df360="" class="fav-button"><i data-v-bd0df360=""><img data-v-bd0df360="" src="/src/assets/star-gold.svg" alt="star-gold"></i></div>"`;
